File: shared/sabre_crypto.cpp
#include "sabre_crypto.h"
#include <openssl/sha.h>
#include <iomanip>
#include <sstream>

std::string calculate_sabre_signature(const std::string& plate,
                                      const std::string& timestamp,
                                      double lat, double lon,
                                      const std::vector<uint8_t>& image_data) {
    SHA256_CTX sha256;
    SHA256_Init(&sha256);

    std::string metadata = plate + "|" + timestamp + "|" + std::to_string(lat) + "|" + std::to_string(lon) + "|";
    SHA256_Update(&sha256, metadata.c_str(), metadata.size());
    SHA256_Update(&sha256, image_data.data(), image_data.size());

    unsigned char hash[SHA256_DIGEST_LENGTH];
    SHA256_Final(hash, &sha256);

    std::stringstream ss;
    for(int i = 0; i < SHA256_DIGEST_LENGTH; i++) {
        ss << std::hex << std::setw(2) << std::setfill('0') << (int)hash[i];
    }
    return ss.str();
}

File: shared/sabre_crypto.py
import hashlib

def generate_sabre_hash(plate: str, timestamp: str, lat: float, lon: float, image_bytes: bytes) -> str:
    """
    Concatenates metadata and image bytes for a unique SHA-256 signature.
    Format: plate|timestamp|lat|lon|image_data
    """
    hasher = hashlib.sha256()
    metadata = f"{plate}|{timestamp}|{lat:.6f}|{lon:.6f}|"
    hasher.update(metadata.encode('utf-8'))
    hasher.update(image_bytes)
    return hasher.hexdigest()
